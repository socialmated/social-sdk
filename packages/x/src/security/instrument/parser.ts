import { gotScraping } from '@social-sdk/client/http';
import { type AnonymousFunctionDeclaration, type FunctionDeclaration, Parser, type ReturnStatement } from 'acorn';
import { simple } from 'acorn-walk';

/**
 * Represents a parsed function containing the source code and its AST node.
 */
export interface ParsedFunction {
  source: string;
  node: FunctionDeclaration;
}

/**
 * Parses the UI metrics JavaScript file to extract the target function that computes UI metrics.
 */
export class UIMetricsJsParser {
  /**
   * The target function that contains the UI metrics logic.
   */
  private targetFunction: FunctionDeclaration | null = null;

  /**
   * Constructs a new instance of the parser and parses the provided script text.
   *
   * @param sourceCode - The JavaScript source code to be parsed.
   */
  constructor(private sourceCode: string) {}

  /**
   * Creates a UIMetricsJsParser by fetching and parsing the UI metrics JavaScript.
   *
   * @returns A promise that resolves to an instance of UIMetricsJsParser.
   */
  public static async create(): Promise<UIMetricsJsParser> {
    const jsFileContent = await gotScraping('https://twitter.com/i/js_inst?c_name=ui_metrics', {
      headers: {
        accept: '*/*',
        referer: 'https://x.com/',
        'sec-fetch-dest': 'script',
        'sec-fetch-mode': 'no-cors',
        'sec-fetch-site': 'cross-site',
      },
    }).text();

    return new UIMetricsJsParser(jsFileContent);
  }

  /**
   * Parses the source code to locate and extract the target function's source as a string.
   *
   * @returns The source code of the target function if found, or null if not found.
   */
  public parseTargetFunction(): ParsedFunction | null {
    const ast = Parser.parse(this.sourceCode, {
      ecmaVersion: 2020,
      sourceType: 'script',
      allowReturnOutsideFunction: true,
    });

    // Walk the AST to find the target function
    simple(ast, {
      FunctionDeclaration: (node) => {
        this.checkFunction(node);
      },
    });

    if (!this.targetFunction) {
      return null;
    }

    return {
      source: this.sourceCode.slice(this.targetFunction.start, this.targetFunction.end),
      node: this.targetFunction,
    };
  }

  /**
   * Checks whether the given function node contains a direct return statement
   * that returns an object with 'rf' and 's' properties.
   *
   * @param node - The function declaration node to check.
   */
  private checkFunction(node: FunctionDeclaration | AnonymousFunctionDeclaration): void {
    // Look for function that has a return statement with an object containing 'rf' and 's' properties
    let hasTargetReturn = false;
    for (const stmt of node.body.body) {
      if (stmt.type === 'ReturnStatement' && this.isTargetReturnStatement(stmt)) {
        hasTargetReturn = true;
        break;
      }
    }

    if (hasTargetReturn) {
      this.targetFunction = node as FunctionDeclaration;
    }
  }

  /**
   * Determines whether a given return statement node returns an object expression
   * containing both 'rf' and 's' properties.
   *
   * @param returnNode - The AST node representing a return statement.
   * @returns `true` if the return statement returns an object with both 'rf' and 's' properties; otherwise, `false`.
   */
  private isTargetReturnStatement(returnNode: ReturnStatement): boolean {
    const returnArg = returnNode.argument;

    // Check if return statement returns an object expression
    if (!returnArg || returnArg.type !== 'ObjectExpression') {
      return false;
    }

    const properties = returnArg.properties;

    // Look for 'rf' and 's' properties
    const hasRfProperty = properties.some(
      (prop) => prop.type === 'Property' && prop.key.type === 'Literal' && prop.key.value === 'rf',
    );
    const hasSProperty = properties.some(
      (prop) => prop.type === 'Property' && prop.key.type === 'Literal' && prop.key.value === 's',
    );

    return hasRfProperty && hasSProperty;
  }
}
