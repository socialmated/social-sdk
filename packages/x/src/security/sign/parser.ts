import { JSDOM, type BinaryData } from 'jsdom';
import { HeaderGenerator } from 'header-generator';
import { gotScraping } from 'got-scraping';

/**
 * Parses X (formerly Twitter) homepage HTML to extract key information required for signing transactions.
 *
 * @remarks
 * - The parser relies on specific HTML patterns present in the X homepage.
 * - If the homepage structure changes, the parser may need to be updated accordingly.
 *
 * @public
 */
class HomeHtmlParser {
  /**
   * Regular expression to extract the ondemand JavaScript file chunk hash from the homepage HTML.
   * The hash is used to construct the URL for fetching the ondemand file.
   */
  private static readonly ON_DEMAND_FILE_REGEX = /['"]ondemand\.s['"]:\s*['"](?<hash>\w*)['"]/;

  /**
   * The document object representing the homepage HTML.
   * This is used to query and manipulate the DOM for extracting information.
   */
  private homeDocument: Document;

  /**
   * Constructs a new parser instance from the homepage HTML.
   * @param htmlText - The HTML content of X's homepage as a string, Buffer, or BinaryData.
   */
  constructor(htmlText: string | Buffer | BinaryData) {
    this.homeDocument = new JSDOM(htmlText).window.document;
  }

  /**
   * Creates a new instance of {@link HomeHtmlParser} by fetching the homepage HTML from x.com.
   *
   * @returns A promise that resolves to a {@link HomeHtmlParser} instance containing the fetched homepage HTML.
   * @throws If the homepage cannot be fetched successfully.
   */
  public static async create(): Promise<HomeHtmlParser> {
    const homeHtml = await gotScraping('https://x.com/home').buffer();
    return new HomeHtmlParser(homeHtml);
  }

  /**
   * Retrieves SVG animation frame elements from the homepage HTML.
   * Frames are identified by element IDs starting with 'loading-x-anim'.
   *
   * @returns Array of SVG frame elements.
   */
  public getAnimationFrames(): SVGElement[] {
    return Array.from(this.homeDocument.querySelectorAll("[id^='loading-x-anim']"));
  }

  /**
   * Locates and fetches the ondemand JavaScript file referenced in the homepage HTML.
   * The file is required for extracting key byte indices used in transaction ID generation.
   *
   * @returns The content of the ondemand file as a string.
   * @throws Error if the ondemand file cannot be found or fetched.
   */
  public findOnDemandJsChunkHash(): string {
    // Serialize the homepage HTML for regex searching.
    const responseStr = this.homeDocument.documentElement.outerHTML;

    // Extract the chunk hash for the ondemand file.
    const onDemandFileMatch = HomeHtmlParser.ON_DEMAND_FILE_REGEX.exec(responseStr);
    const chunkHash = onDemandFileMatch?.groups?.['hash'];
    if (!chunkHash) {
      throw new Error("Couldn't extract ondemand file chunk hash from the page source");
    }

    return chunkHash;
  }
  /**
   * Extracts the verification key from the homepage HTML.
   * The verification key is stored in a meta tag with the name 'twitter-site-verification'.
   *
   * @returns The verification key as a string.
   * @throws Error if the verification key cannot be found.
   */
  public getVerificationKey(): string | null {
    // Find the meta tag containing the verification key.
    const element = this.homeDocument.querySelector("[name='twitter-site-verification']");
    if (!element) {
      throw new Error("Couldn't find meta tag with name 'twitter-site-verification'");
    }

    return element.getAttribute('content');
  }
}

/**
 * Parses a remote ondemand JavaScript file to extract cryptographic key byte indices.
 *
 * @remarks
 * - The ondemand JavaScript file is fetched from a Twitter CDN using a provided chunk hash.
 * - The indices are extracted using a regular expression and are expected to change frequently.
 *
 * @public
 */
class OnDemandJsParser {
  /**
   * Regular expression to extract key byte indices from the ondemand JavaScript file.
   * The indices are used for cryptographic key derivation.
   */
  private static readonly INDICES_REGEX = /\(\w\[(?<index>\d{1,2})\],\s*16\)/g;

  /**
   * Constructs a new parser instance from the ondemand JavaScript file content.
   * @param sourceCode - The JavaScript source code of the ondemand file as a string.
   */
  constructor(private sourceCode: string) {}

  /**
   * Creates an instance of `OnDemandJsParser` by fetching and parsing a remote JavaScript file
   * identified by the provided `chunkHash`.
   *
   * @param chunkHash - The hash string used to construct the URL for the ondemand JavaScript file.
   * @returns A promise that resolves to an `OnDemandJsParser` instance containing the fetched JavaScript content.
   * @throws If the fetch request fails or the response is not OK.
   */
  public static async create(chunkHash: string): Promise<OnDemandJsParser> {
    const jsResponse = await fetch(`https://abs.twimg.com/responsive-web/client-web/ondemand.s.${chunkHash}a.js`, {
      // Requests will be blocked if user-agent is unknown.
      headers: new HeaderGenerator().getHeaders(),
    });
    if (!jsResponse.ok) {
      throw new Error(`Failed to fetch ondemand file: ${jsResponse.statusText}`);
    }

    const jsContent = await jsResponse.text();

    return new OnDemandJsParser(jsContent);
  }

  /**
   * Extracts key byte indices from the ondemand file content.
   * These indices are used for cryptographic key derivation.
   * Keep in mind that these indices changes frequently.
   *
   * @param onDemandFile - The text content of the ondemand file. If omitted, uses the last fetched file.
   * @returns Tuple of [rowIndex, keyByteIndices], where keyByteIndices is an array of three numbers.
   * @throws Error if the key byte indices cannot be extracted.
   */
  public getIndices(): [number, number, number, number] {
    const keyByteIndices: number[] = [];

    // Extract all index matches using the regex.
    const matches = Array.from(this.sourceCode.matchAll(OnDemandJsParser.INDICES_REGEX));
    for (const match of matches) {
      const index = match.groups?.['index'];
      if (!index) {
        throw new Error("Couldn't extract index from regex match");
      }
      keyByteIndices.push(parseInt(index, 10));
    }

    return keyByteIndices as [number, number, number, number];
  }
}

export { HomeHtmlParser, OnDemandJsParser };
