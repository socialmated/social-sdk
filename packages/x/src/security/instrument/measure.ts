import { JSDOM } from 'jsdom';
import { type ParsedFunction, UIMetricsJsParser } from './parser.js';

/**
 * Measures UI metrics by evaluating a specific function in a simulated browser environment.
 *
 * @example
 * ```typescript
 * const measurer = new UIMetricsMeasurer();
 * const metrics = await measurer.measure();
 * console.log(metrics.rf, metrics.s);
 * ```
 */
export class UIMetricsMeasurer {
  /**
   * The JSDOM instance used to simulate a browser environment for UI metrics measurement.
   */
  private dom: JSDOM;

  /**
   * The parsed UI metrics function extracted from the JavaScript file.
   */
  private uiMetricsFunction?: ParsedFunction;

  /**
   * Initializes the class with a JSDOM environment simulating a browser.
   */
  constructor() {
    this.dom = new JSDOM(`<body></body>`, {
      url: 'https://twitter.com',
      runScripts: 'dangerously',
      pretendToBeVisual: true,
    });
  }

  /**
   * Measures UI metrics by evaluating the parsed function in the JSDOM environment.
   *
   * @returns A promise that resolves to an object containing the UI metrics results.
   */
  public async measure(): Promise<{
    rf: Record<string, number>;
    s: string;
  }> {
    if (!this.uiMetricsFunction) {
      await this.init();
    }
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- expected to be defined after init
    const uiMetricsFunction = this.uiMetricsFunction!;

    // Evaluate the function in the JSDOM environment
    this.dom.window.eval(uiMetricsFunction.source);
    const result = this.dom.window.eval(`${uiMetricsFunction.node.id.name}()`);

    return result as {
      rf: Record<string, number>;
      s: string;
    };
  }

  /**
   * Initializes the UI metrics parser and extracts the target function.
   *
   * @returns A promise that resolves when initialization is complete.
   * @throws If the target function cannot be found in the UI metrics JavaScript file.
   */
  private async init(): Promise<void> {
    const parser = await UIMetricsJsParser.create();

    // Get the target function from the UI metrics JavaScript file
    const targetFunction = parser.parseTargetFunction();
    if (!targetFunction) {
      throw new Error('Target function not found in UI metrics JavaScript file');
    }
    this.uiMetricsFunction = targetFunction;
  }
}
