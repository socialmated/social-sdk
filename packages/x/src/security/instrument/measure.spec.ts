import { readFileSync } from 'node:fs';
import path from 'node:path';
import { UIMetricsMeasurer } from './measure.js';
import { UIMetricsJsParser } from './parser.js';

describe(UIMetricsMeasurer, () => {
  const jsInstFile = readFileSync(path.join(import.meta.dirname, '__fixtures__', 'js_inst.js'), 'utf-8');

  let measurer: UIMetricsMeasurer;

  beforeEach(() => {
    vi.spyOn(UIMetricsJsParser, 'create').mockResolvedValue(new UIMetricsJsParser(jsInstFile));

    measurer = new UIMetricsMeasurer();
  });

  describe(UIMetricsMeasurer.prototype.measure, () => {
    it('should measure the ui metrics', async () => {
      const result = await measurer.measure();

      expect(result).toMatchSnapshot();
    });
  });
});
