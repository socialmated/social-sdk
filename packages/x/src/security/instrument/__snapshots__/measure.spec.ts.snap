// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UIMetricsMeasurer > measure > should measure the ui metrics 1`] = `
{
  "rf": {
    "a1271ddd5495959ff47b6b0ede71053633d2d6b42ad2f5a65b35bc01a8f25ce6": -66,
    "a5bb6874e28ab18bdce885228f6e98e4a7efcf3c3bb21509f9859e1f712d5281": -66,
    "a6fdd0ffaf8cf5b5a0f6597a6b2d6ad4409de03d95995cfeacfcd95f637fe294": 18,
    "afce4f5d264e1ce474bdb74712aef91022be4b54ac665bb4c64c4a05d3dade31": 65,
  },
  "s": "xHGdrB7Sr3_W7HgSzB6Xrwl6IKBB7nEg2tJZPZqTy9Dvgz8Hs-LYRoYRjXhvdipXvBZdVC33_Jgx4A_34ZiMRii5sGNJHA4-KUmee_8TA2sehD5oqlkpgRkNgexmGkrhnGBBj5aRr-yldWfWZEagUZOJVcqIxLXSSag5eSqncsDS7Tpo-8aLD5c4mmDwBHKX0w-o-rNshoZ-VaDyIYQYCWU9CfTaex10noBGRZCF2snhxSGWMbIk44NiLdaqhxXqN9IeS0GP4bDcdByZt-EdUixAigq9KQI_hKV_1sg880gdXaPcQGJWI7X33Dbnzc7oGQjKBS60-lIysH2kHdhdaAAAAZaZ4FKT",
}
`;
